### PR TITLE
Deprecate _source.mode in mappings

### PR DESCRIPTION
**Pull request by dnhatn**
This change deprecates _source.mode in mappings, replacing it with the `index.mapping.source.mode` index setting.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 1]
```diff
@@ -1411,6 +1411,17 @@ private static void updateIndexMappingsAndBuildSortOrder(
         }
     }
 
+    private static void validateSourceMode(MapperService mapperService) {
+        final IndexSettings indexSettings = mapperService.getIndexSettings();
+        if (indexSettings != null
+            && indexSettings.getIndexVersionCreated().onOrAfter(IndexVersions.DEPRECATE_SOURCE_MODE_MAPPER)
+            && SourceFieldMapper.isSynthetic(indexSettings) == false
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:1418]: I think we need to check whether `index.mapping.source.mode` index setting has not be defined here? Other we only check whether this setting has been set to `synthetic`?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 2]
```diff
@@ -1411,6 +1411,17 @@ private static void updateIndexMappingsAndBuildSortOrder(
...
+            && mapperService.documentMapper() != null
+            && mapperService.documentMapper().sourceMapper().isSynthetic()) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:1420]: If we check whether `index.mapping.source.mode` has not be defined, then I think ` mapperService.documentMapper().sourceMapper().isSynthetic()` can return true if index.mode=logsdb|time_series and no _source.mode has been defined?

**Review by martijnvg — COMMENTED** [review]
Thanks Nhat, I left questions are the deprecation logic in `MetadataCreateIndexService`.

I wonder whether the following is perhaps a better mechanism to emit depreciation warning in case _source.mode is used: 

```patch
Subject: [PATCH] Adjust SyntheticSourceLicenseService 

Allow gold and platinum license to use synthetic source for a limited time.
If the start time of a license if before a cut off date, then gold and platinum licenses will not fallback to stored source if synthetic source is used.
---
Index: server/src/main/java/org/elasticsearch/index/mapper/MappingParser.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/server/src/main/java/org/elasticsearch/index/mapper/MappingParser.java b/server/src/main/java/org/elasticsearch/index/mapper/MappingParser.java
--- a/server/src/main/java/org/elasticsearch/index/mapper/MappingParser.java	(revision bd091d3d96b33adb4121f980dc4f9f7a2a87b043)
+++ b/server/src/main/java/org/elasticsearch/index/mapper/MappingParser.java	(date 1731573325979)
@@ -10,6 +10,8 @@
 package org.elasticsearch.index.mapper;
 
 import org.elasticsearch.common.compress.CompressedXContent;
+import org.elasticsearch.common.logging.DeprecationCategory;
+import org.elasticsearch.common.logging.DeprecationLogger;
 import org.elasticsearch.common.xcontent.XContentHelper;
 import org.elasticsearch.core.Nullable;
 import org.elasticsearch.index.mapper.MapperService.MergeReason;
@@ -27,6 +29,9 @@
  * Parser for {@link Mapping} provided in {@link CompressedXContent} format
  */
 public final class MappingParser {
+
+    private static final DeprecationLogger DEPRECATION_LOGGER = DeprecationLogger.getLogger(MappingParser.class);
+
     private final Supplier<MappingParserContext> mappingParserContextSupplier;
     private final Supplier<Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper>> metadataMappersSupplier;
     private final Map<String, MetadataFieldMapper.TypeParser> metadataMapperParsers;
@@ -144,6 +149,14 @@
                 }
                 @SuppressWarnings("unchecked")
                 Map<String, Object> fieldNodeMap = (Map<String, Object>) fieldNode;
+
+                // iirc after parsing fieldNodeMap, keys are removed
+                if (reason == MergeReason.MAPPING_UPDATE && SourceFieldMapper.CONTENT_TYPE.equals(fieldName)) {
+                    if (fieldNodeMap.containsKey("mode")) {
+                        DEPRECATION_LOGGER.critical(DeprecationCategory.MAPPINGS, "mapping_source_mode", SourceFieldMapper.DEPRECATION_WARNING);
+                    }
+                }
+
                 MetadataFieldMapper metadataFieldMapper = typeParser.parse(fieldName, fieldNodeMap, mappingParserContext).build();
                 metadataMappers.put(metadataFieldMapper.getClass(), metadataFieldMapper);
                 assert fieldNodeMap.isEmpty();

```

I think this also returns warning when mapping gets updated, maybe that isn't too bad? But this should only return a warning when _source.mode attribute is really used.

**Comment by dnhatn** [comment]
@martijnvg I think it's ready. Can you please take another look? Thanks!

**Review by martijnvg — APPROVED** [review]
Left a question and minor comments. LGTM otherwise.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/compat/compat/RestCompatTestTransformTask.java [hunk 3]
```diff
@@ -137,7 +137,7 @@ public void skipTest(String fullTestName, String reason) {
         // However, the folder can be arbitrarily nest so, a == a1/a2/a3, and the test name can include forward slashes, so c == c1/c2/c3
         // So we also need to support a1/a2/a3/b/c1/c2/c3
 
-        String[] testParts = fullTestName.split("/");
+        String[] testParts = fullTestName.split("/", 3);
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/compat/compat/RestCompatTestTransformTask.java:140]: To be able to skip `logsdb/20_source_mapping/include/exclude is supported with stored _source`, which contains `/` in its name, in the rest-compat tests.

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/DoSection.java [hunk 4]
```diff
@@ -502,6 +503,8 @@ public void checkWarningHeaders(final List<String> warningHeaders, String testPa
             }
         }
 
+        unexpected.removeIf(s -> s.endsWith(SourceFieldMapper.DEPRECATION_WARNING + "\""));
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/DoSection.java:506]: Several of them. I had planned to isolate this check in a follow-up.

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java [hunk 5]
```diff
@@ -1840,8 +1842,9 @@ public static CreateIndexResponse createIndex(RestClient client, String name, Se
 
         if (settings != null && settings.getAsBoolean(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true) == false) {
             expectSoftDeletesWarning(request, name);
+        } else if (isSyntheticSourceConfiguredInMapping(mapping)) {
+            request.setOptions(expectVersionSpecificWarnings(v -> v.compatible(SourceFieldMapper.DEPRECATION_WARNING)));
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:1846]: Same here -  several of them. I had planned to isolate this check in a follow-up."

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/compat/compat/RestCompatTestTransformTask.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (martijnvg) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/test/rest/compat/compat/RestCompatTestTransformTask.java:140]: Ok, makes sense. Maybe add a comment here explaining that only three parts are needed?

**Review by mark-vieira — APPROVED** [review]

File: x-pack/plugin/logsdb/qa/with-basic/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsdbWithBasicRestIT.java [hunk 6]
```diff
@@ -85,14 +85,20 @@ public void testLogsdbIndexGetsStoredSource() throws IOException {
...
+                "mappings": {
+                    "_source": {
+                        "mode": "synthetic"
+                    }
                 }
             }
-            """;
-        createIndex(index, Settings.builder().put("index.mode", "logsdb").build(), mapping);
+            """);
+        assertOK(client().performRequest(request));
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply x-pack/plugin/logsdb/qa/with-basic/src/javaRestTest/java/org/elasticsearch/xpack/logsdb/LogsdbWithBasicRestIT.java]: Yes, I have reverted this.

**Comment by dnhatn** [comment]
@martijnvg @mark-vieira Thanks for reviewing.

**Pull request merged**